src/analyses/dirty.h

```diff
@@ -17,42 +17,74 @@ Date: March 2013
 #include <unordered_set>
 
 #include <util/std_expr.h>
+#include <util/invariant.h>
 #include <goto-programs/goto_functions.h>
 
 class dirtyt
 {
+private:
+  void die_if_uninitialized() const
+  {
+    INVARIANT(initialized,
+        "Uninitialized dirtyt. This dirtyt was constructed using the default "
+        "constructor and not subsequently initialized using "
+        "dirtyt::build().");
+  }
+
 public:
+  bool initialized;
   typedef std::unordered_set<irep_idt, irep_id_hash> id_sett;
   typedef goto_functionst::goto_functiont goto_functiont;
 
-  explicit dirtyt(const goto_functiont &goto_function)
+  dirtyt():
+    initialized(false)
+  {
+  }
+
+  explicit dirtyt(const goto_functiont &goto_function):
+    initialized(false)
   {
     build(goto_function);
+    initialized=true;
   }
 
-  explicit dirtyt(const goto_functionst &goto_functions)
+  explicit dirtyt(const goto_functionst &goto_functions):
+    initialized(false)
   {
-    forall_goto_functions(it, goto_functions)
-      build(it->second);
+    build(goto_functions);
+    // build(g_funs) responsible for setting initialized to true, since
+    // it is public and can be called independently
   }
 
   void output(std::ostream &out) const;
 
   bool operator()(const irep_idt &id) const
   {
+    die_if_uninitialized();
     return dirty.find(id)!=dirty.end();
   }
 
   bool operator()(const symbol_exprt &expr) const
   {
+    die_if_uninitialized();
     return operator()(expr.get_identifier());
   }
 
   const id_sett &get_dirty_ids() const
   {
+    die_if_uninitialized();
     return dirty;
   }
 
+  void build(const goto_functionst &goto_functions)
+  {
+    INVARIANT(!initialized,
```

Comment:
I think this should be a PRECONDITION because it's a condition on the structure which must hold for this to be called,